cpp/src/arrow/compute/kernels/isin-test.cc

```diff
@@ -0,0 +1,391 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <cstdint>
+#include <cstdio>
+#include <functional>
+#include <locale>
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/array.h"
+#include "arrow/buffer.h"
+#include "arrow/compute/context.h"
+#include "arrow/compute/kernel.h"
+#include "arrow/compute/kernels/isin.h"
+#include "arrow/compute/kernels/util-internal.h"
+#include "arrow/compute/test-util.h"
+#include "arrow/memory_pool.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/testing/gtest_common.h"
+#include "arrow/testing/util.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/decimal.h"
+
+namespace arrow {
+namespace compute {
+
+// ----------------------------------------------------------------------
+// IsIn tests
+
+template <typename Type, typename T = typename TypeTraits<Type>::c_type>
+void CheckIsIn(FunctionContext* ctx, const std::shared_ptr<DataType>& type,
+               const std::vector<T>& in_values, const std::vector<bool>& in_is_valid,
+               const std::vector<T>& member_set_values,
+               const std::vector<bool>& member_set_is_valid,
+               const std::vector<bool>& out_values,
+               const std::vector<bool>& out_is_valid) {
+  std::shared_ptr<Array> input = _MakeArray<Type, T>(type, in_values, in_is_valid);
+  std::shared_ptr<Array> member_set =
+      _MakeArray<Type, T>(type, member_set_values, member_set_is_valid);
+  std::shared_ptr<Array> expected =
+      _MakeArray<BooleanType, bool>(boolean(), out_values, out_is_valid);
+
+  Datum datum_out;
+  ASSERT_OK(IsIn(ctx, input, member_set, &datum_out));
+  std::shared_ptr<Array> result = datum_out.make_array();
+  ASSERT_ARRAYS_EQUAL(*expected, *result);
+}
+
+class TestIsInKernel : public ComputeFixture, public TestBase {};
+
+template <typename Type>
+class TestIsInKernelPrimitive : public ComputeFixture, public TestBase {};
+
+typedef ::testing::Types<Int8Type, UInt8Type, Int16Type, UInt16Type, Int32Type,
+                         UInt32Type, Int64Type, UInt64Type, FloatType, DoubleType,
+                         Date32Type, Date64Type>
+    PrimitiveDictionaries;
+
+TYPED_TEST_CASE(TestIsInKernelPrimitive, PrimitiveDictionaries);
+
+TYPED_TEST(TestIsInKernelPrimitive, IsIn) {
+  using T = typename TypeParam::c_type;
+  auto type = TypeTraits<TypeParam>::type_singleton();
+  CheckIsIn<TypeParam, T>(&this->ctx_, type, {2, 1, 2, 1, 2, 3},
+                          {true, false, true, true, true, true}, {2, 1, 2, 1, 2, 3},
+                          {true, false, true, true, true, true},
+                          {true, true, true, true, true, true}, {});
+}
```

Comment:
I think we want tests for null/not null cases here as well?

Reply:
Added the null/not null/no match cases.